### DISCUSSION TITLE
TDR or Impedance mismatch analysis

### DISCUSSION THREAD

**Discussion by lowpon** [thliebig/openEMS-Project#177] [Q&A Usage]
Hi all,
First I'd like to thank M. Liebig for this amazing tool and M. Jagos for its FreeCAD to openEMS very helpful macro.
I started few weeks ago with openEMS and got recently my first results.
Based on tutorials, as shown below, it was easy to plot s-parameters in the frequency domain or voltage in the time domain, but how could I plot the impedance in the time domain now ?
This would be a TDR analysis, showing impedance mismatch along the transmission line.
Thanks for your help

![image](https://github.com/thliebig/openEMS-Project/assets/156537075/b2173a92-7cc2-4c52-b848-0f8c0b3a23e2)

![image](https://github.com/thliebig/openEMS-Project/assets/156537075/b445d59f-6652-4cfb-9ffb-749c84f46ab1)

**Comment by LubomirJagos** [comment]
Hi, thanks for using my addon, what do you mean plot impedance in time domain? Impedance is not changing over time, there are voltage and current captured in time domain and impedance is always calculated as U/I

**Reply by thliebig** [reply]
It's not about time. Its about location. The idea is to send a signal and from the reflections determine how the impedance changes a long the length of the e.g. transmission line. Most importantly to e.g. find damages on a TL.

See [here](https://en.wikipedia.org/wiki/Time-domain_reflectometer).

I'm not aware that anybody has done this with openEMS thus far and I'm not familiar with the details of TDR. 
I only know that the FDTD method can be used for it. 
But I fear that you have to implement the necessary processing yourself.

**Reply by thliebig** [reply]
But on reading your description again I would agree that calling it "Impedance in Time-Domain" is not really accurate and can indeed be a bit misleading...

**Reply by lowpon** [reply]
You're right, I have modified the title.
I'll come back soon with first results following a step excitation, but I'm not sure to be able to define the risetime corresponding to my signal speed. For sure, Ohm's law will be convenient.

**Comment by VolkerMuehlhaus** [comment]
@lowpon Just in case someone wants to post example code: Do you use openEMS with Python or Matlab?

I have done TDR analysis using other EM tools [(appnote)](https://muehlhaus.com/support/empire-appnotes/rfic-pcb-tdr) and can provide some basic hints:
My preferred excitation for line discontinuity analysis is a simple step transition: voltage off a short period of time, then jump to voltage on and hold that voltage. 
Simulation timesteps are used here to limit the FDTD simulation time: we only need to capture a short period of time after the initial pulse, it is not necessary to run further until energy threshold is reached.
For evaluation I look at reflected voltage, step down means the impedance at that point reached by the pulse is too low (or capacitive) and step up means impdance is too high (or inductive). 
To get started, I recommend to build some simple test structures line a line with **known** discontinuities, to get familar with the resulting patterns in reflected signal in time domain.

What I described above is based on reflected voltage for a single step, and things get complicated when you reach a time where multiple reflections between multiple discontinuities show up. But as you can see from the TDR appnote that I linked above, it is already quite useful.

Getting true impedance over time (or distance) without that multi-reflection issue is more complicated. One solution that I use at work is the TDR testbench in Keysight ADS, which removes these multi-reflection artefacts. I don't know how to perform that task in openEMS.

**Reply by lowpon** [reply]
Nice appnote thanks !
I use Octave/Matlab.

**Reply by VolkerMuehlhaus** [reply]
Here is a quick & dirty testbench that I drafted using Python syntax, you should be able to reproduce that in Matlab syntax.

Geometry is a 100mm long line with ~50 Ohm on FR4 (line w=2.7mm on FR4 epsr=4.4 h=1.5mm), with a 20mm long section of ~35 Ohm line as discontinuity. The model uses step excitation. 

edit: model updated, fixed error in max_cellsize calculation
[tdr_line_discont2.zip](https://github.com/thliebig/openEMS-Project/files/13939630/tdr_line_discont2.zip)

![discont_geometry](https://github.com/thliebig/openEMS-Project/assets/47922981/496e7130-3bc2-4105-aa27-154c03daf393)

To get started, we can look at normalized voltages at ports 1 and 2, to identify the time of flight along the line: approximately 700pS for 100mm line length. In air, we would expect 100mm / 3E8m/s = 333ps, but the microstrip on FR4 has more group delay due to its permittivity.

For reflected voltage, we see a drop around 400ps. This is the path to discontinuity and back, so the single path distance can be estimated as 200pS / (700ps/100mm) = 28.6mm, which is in reasonable agreement with the actual start of the discontinuity at 30mm.

![discont_voltages](https://github.com/thliebig/openEMS-Project/assets/47922981/f83e0d05-1a2b-42ed-bcb3-3e87e7eed6cd)

Magnitude and sign of the reflected signal provide the reflection magnitude, and the negative reflection factor indicates an impedance less than 50 Ohm here. We can estimate Z = Z0 * (1+r)/(1-r)

This is only a rough draft to get the discussion going. You can see a ringing in port 2 voltaged, for example. Overall the method suggested below by biergazi looks cleaner, but it requires a fully converged model with all paths excited to get the full [S] matrix. The timestep TDR model shown here only runs a few timesteps, so it is quite fast and might be sufficient to get started.

![discont_estimated_Z](https://github.com/thliebig/openEMS-Project/assets/47922981/fb983c20-65eb-4209-a4a0-db9e34474bd8)

**Reply by VolkerMuehlhaus** [reply]
For testing the method suggested by biergaizi, here is the S2P data for the model above, obtained from full 2-port simulation up to 10 GHz with gaussian pulse.

[discont_sparam_2port.zip](https://github.com/thliebig/openEMS-Project/files/13940628/discont_sparam_2port.zip)

result:

![tdr_from_sparam](https://github.com/thliebig/openEMS-Project/assets/47922981/f725cf86-a9c3-4e95-baf7-b32c3762c119)

**Reply by lowpon** [reply]
Your test case will be helpful to train with Octave and try to reproduce. Thanks.

**Reply by lowpon** [reply]
Here is what I get with step excitation.
My meshing is not exactly the same as yours but seems working
![PCB](https://github.com/thliebig/openEMS-Project/assets/156537075/9e711940-e4bb-4df7-91c2-0e17c8f3b94a)

Voltages look similar even with lower levels

![PCB_Voltage](https://github.com/thliebig/openEMS-Project/assets/156537075/b985076b-94ca-4b05-bf08-4c63f467696c)

But impedance is quite strange

![PCB_Impedance](https://github.com/thliebig/openEMS-Project/assets/156537075/cf8f1e6a-0f58-4ff2-8ff3-6d940c45dffc)

**Reply by VolkerMuehlhaus** [reply]
Note that my voltages are normalized to the incident voltege, that is the starting value is 1.0
What equation did you use for calculating line impedance?

**Reply by lowpon** [reply]
I put the files here
https://github.com/lowpon/openEMS-TDR/blob/5d9efc6081cf1570a4bd9fb26073041b4f66f55d/PCB_draw_Sparam.m

**Reply by VolkerMuehlhaus** [reply]
@lowpon 

Your impedance calculation is wrong, and there are better method to access the port voltages. 

Access port voltage and current in time domain, instead of reading the files:
%   port.ut.tot     total voltage (time-domain)
%   port.ut.time    voltage time vector
%   port.it.tot     total current (time-domain)
%   port.it.time    current time vector

As said, your simple calculation of Z=U/I is wrong, because that is the not the calculated line impedance. You can use the method shown in my Python example to calculate incident and reflected waves, and calculate line impedance from there:

```
portZ = Z0
U1_inc = 0.5 * (U1 + I1 * portZ)
U1_ref = U1 - U1_inc
U2_inc = 0.5 * (U2 + I2 * portZ)
U2_ref = U2 - U2_inc

# estimate impedance magnitude
r = U1_ref / U1_inc
Z = Z0 * (1+r)/(1-r)
```

I don't like working with the openEMS Matlab interface, and your coode is a bit overloaded, so I didn't fix your code, but hope you can make the changes yourself.

**Reply by lowpon** [reply]
Thank you @VolkerMuehlhaus 
Here are the last results following your reply.
I still don't know why impedance seems to be referenced to 100ohm and voltage levels different from yours

![image](https://github.com/thliebig/openEMS-Project/assets/156537075/6936ae8d-cd2d-4c59-a231-8c90db211a9a)

![image](https://github.com/thliebig/openEMS-Project/assets/156537075/68e7bc1e-dd6d-431c-b124-53b659a3ef19)

Below, an extract but the complete code is available here
https://github.com/lowpon/openEMS-TDR/blob/76ad73a71ddd2fb86f5d6afb3dc4f3e16e1b27fb/PCB_TDR

```octave
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% POST-PROCESSING AND TIME DOMAIN PLOT GENERATION
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

freq = linspace( fstart, fstop, numfreq );
port{1} = calcPort( port{1}, Sim_Path, freq, 'RefImpedance', Z0);
port{2} = calcPort( port{2}, Sim_Path, freq, 'RefImpedance', Z0);

% port.ut.tot total voltage (time-domain)
% port.it.tot total current (time-domain)
% port.ut.time voltage time vector
% port.it.time current time vector (same as voltage)

U1 = port{1}.ut.tot;
U2 = port{2}.ut.tot;
I1 = port{1}.it.tot;
I2 = port{2}.it.tot;
t = port{1}.ut.time;

U1_inc = 0.5 * (U1 + I1.* Z0);
U1_ref = U1 - U1_inc;
U2_inc = 0.5 * (U2 + I2.* Z0);
U2_ref = U2 - U2_inc;

figure
plot( t, U1./ U1_inc, 'k-', 'Linewidth', 2 );
hold on;
plot( t, U2./ U1_inc, 'r--', 'Linewidth', 2 );
hold on;
plot( t, U1_ref./ U1_inc, 'b--', 'Linewidth', 2 );
%ylim([0 2]);
xlim([0 6e-10]);
grid on;
grid minor on;
legend( 'U1', 'U2', 'U1 ref' );
ylabel('Port voltages (normalized to incident)');
xlabel('Time (s)');
savefig('voltages.png');

%% estimate impedance magnitude
r = U1_ref./ U1_inc;
Z = Z0 * (1+r)./(1-r);

figure
plot( t, Z, 'b-', 'Linewidth', 2 );
ylim([60 120]);
xlim([0 6e-10]);

grid on;
grid minor on;

title( 'Impedance vs Time' );
xlabel( 'Time (s)' );
ylabel( 'Impedance (ohm)' );
legend( 'Z' );
```

**Reply by VolkerMuehlhaus** [reply]
We can't reproduce your issue because you didn't share the model, only the plotting part. You should check if port impedances are consistent between model code and plotting code.

**Reply by lowpon** [reply]
The Freecad model, XML, STL, model code, plotting code and configuration file for @LubomirJagos plugin are now available here
https://github.com/lowpon/openEMS-TDR/blob/0a2b433749edc7274c32ee651730e4f142264606/PCB.FCStd

**Reply by VolkerMuehlhaus** [reply]
Looking at your port 2 voltage above, that also looks unexpected. I guess the issue is more than just port impedances.

Your XML isn't easy to debug, because it links the STL files in your file system locations. I give up here, this turns out to be very time consuming with your workflow. Hopefully Lubomir can help you, he is more familiar with the workflow that you used.

Best regards
Volker

**Reply by lowpon** [reply]
Thanks for your time Volker
Cross fingers with @LubomirJagos
Regards

**Comment by biergaizi** [comment]
> Based on tutorials, as shown below, it was easy to plot s-parameters in the frequency domain or voltage in the time domain, but how could I plot the impedance in the time domain now ?

TL;DR: You can calculate TDR responses from S-parameters using the powerful Python software library `scikit-rf` with 1 line of code. I've used it with both measured data and openEMS simulation results and it works perfectly (of course, within the limitation of calculating Inverse Fourier Transform from S-parameters). Read the documentation of [scikit-rf](https://scikit-rf.readthedocs.io/en/latest/examples/networktheory/Time%20domain%20reflectometry%2C%20measurement%20vs%20simulation.html) for more information.

Using S-parameters for time-domain simulations like these is supported by almost all commercial and even some free RF circuit simulators. This is the main reason why I think using custom time-domain excitation within openEMS is usually unnecessary. I'd like to think of openEMS as a virtual network analyzer, once you've obtained the frequency-domain S-parameters, one can do nearly every kind of further analysis imaginable using standard RF circuit tools at its downstream (but do beware causality and passivity). Of course, trying to just openEMS as "just" a VNA can be inefficient - you're simulating in time domain and converting it to frequency domain before converting it to time domain again (with an appropriate simulation setup, one can avoid the extra work), but the advantage is that one can use tools one is already familiar with. 

---

By taking an Inverse Fourier Transform, it's possible to convert S11 measurements from a VNA to its equivalent time-domain response as measured by a TDR. Previously, this capability is often exclusive in expensive simulation packages or in the firmware of expensive VNAs. Now, with the power of [scikit-rf](http://scikit-rf.org/), a free and open source Python software library for RF/microwave modeling and analysis, it's accessible to everyone with basic programming knowledge.

Here's a demo from my recent experiment (this is only meant to be a quick example, the measurement is not fully calibrated and the parameters used for the calculations are not optimized, so please don't complain too much about the data).

To evaluate the impedance discontinuity and signal integrity impact caused by a 1206 resistor footprint, I fabricated a test fixture shown in the picture below. 

[![PCB test board][1]][1]

[![resistor footprint, with ground plane partially removed][2]][2]

It's a 4-layer circuit board with three solid ground planes. Two through-hole SMA connectors are located on both ends, with a 50-ohm microstrip in between. At the center is the Device Under Test, a 0-ohm, 1206 resistor. Also, to compensate for the excess capacitance caused by the large resistor pads, the ground plane under the resistor body is partially removed. 

After I've received the board, I measured its S11 from 10 MHz to 4 GHz using a Vector Network Analyzer. It shows the return loss is good up to 1.2 GHz. But is it the full picture?

[![Measured S11 data][3]][3]

Using the scikit-rf software library and the following Python code, we can transform the data from frequency domain to time domain.

    import sys
    import skrf
    import matplotlib.pyplot as plt
    skrf.stylely()

    network = skrf.Network(sys.argv[1])
    network_dc = network.extrapolate_to_dc(kind='linear')

    plt.figure()
    plt.title("Time Domain Reflectometry")
    network_dc.s11.plot_z_time_step(window='hamming', label="impedance")
    plt.xlim((-0.5, 1.5))

    plt.tight_layout()
    plt.show()

[![TDR plot][4]][4]

This calculated TDR plot immediately reveals that the test fixture was poorly designed with non-ideal SMA-to-microstrip transitions. When the signal hits the first SMA connector, the instantaneous impedance drops to 45 ohms due to a capacitive discontinuity, then it hits the 50-ohm trace, hits the device-under-test, hits the 50-ohm trace again, and finally hits the second connector.

In fact, after de-embedding, it turned out that the resistor footprint only has 20 dB return loss at 800 MHz, not 1.2 GHz. The bad SMA connector footprints and the resistor footprint actually compensated each other. It would be difficult to spot this problem using the Smith Chart, on the other hand, it's obvious from the computed TDR response. 

You can find more information from scikit-rf's documentation, [Time domain reflectometry, measurement vs simulation](https://scikit-rf.readthedocs.io/en/latest/examples/networktheory/Time%20domain%20reflectometry%2C%20measurement%20vs%20simulation.html). scikit-rf also supports time-gating, it's possible to use time-gating as a simple way to de-embed the connectors from the DUT by deleting the reflections from the time domain, see [Time Domain and Gating](https://scikit-rf.readthedocs.io/en/latest/examples/networktheory/Time%20Domain.html).

It's also worth pointing out that in principle, all calibration and de-embedding techniques developed for VNA measurements can also be applied to simulation data, such as SOLT, TRL, or IEEE P370 - all of these are implemented in scikit-rf, just pass the S-parameters to scikit-rf and [call the APIs](https://scikit-rf.readthedocs.io/en/latest/api/calibration/index.html#two-port), and you get calibrated results. In my opinion, these are especially useful as VNA calibration is a solved problem with mature solutions, so one can use pre-existing tools to do that. For example, if the signal launch at the port is imperfect, one can create a virtual SOLT test fixture in openEMS and run those simulations as "control", then those mismatches can be removed in later analysis.

  [1]: https://i.stack.imgur.com/4h34U.png
  [2]: https://i.stack.imgur.com/5SJWH.png
  [3]: https://i.stack.imgur.com/ppJph.png
  [4]: https://i.stack.imgur.com/buWnG.png

**Reply by VolkerMuehlhaus** [reply]
This is very useful, thank you!

**Reply by lowpon** [reply]
Even if I'd rather work with the native time domain data, I really appreciate your scikit-rf proposal where I discover your huge and inspiring contribution. Thanks @biergaizi
